[jit] make QualifiedName a value type

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19575 [jit] Use QualifiedName for classes
* **#19626 [jit] make QualifiedName a value type**
* #19566 [jit] separate QualifiedName into its own file

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

Differential Revision: [D15049582](https://our.internmc.facebook.com/intern/diff/D15049582)

suo committed:
[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

zdevito reviewed (commented):
In person: let's anticipate a future when someone will use this class where the names might not be valid identifiers and may contain the standard delimiter. We do not have this case now, but we also do not want to have to make a third qualified name class.

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

eellison commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -1,91 +1,96 @@
-    if (prefix_ == nullptr) {
-      ss << name_;
-      return ss;
+  char delimiter_ = '.';
```
this could be const

suo replied:
making members const deletes their copy constructor, which is weird for something that should be a value type. Having it be private and all methods const accomplishes much the same goal

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head

eellison commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -1,91 +1,96 @@
-  static QualifiedNamePtr createFromDotted(const std::string& name) {
+  // `name` can be a dotted string, like "foo.bar.baz", or just a bare name.
+  explicit QualifiedName(std::string name) {
+    AT_ASSERT(!name.empty());
```
nit: this check is unnecessary, and the one below has a better msg

eellison commented on aten/src/ATen/core/qualified_name.h:
```diff
@@ -1,91 +1,96 @@
-        QualifiedNamePtr(), std::move(name));
-  }
+// Represents a name of the form "foo.bar.baz"
+struct QualifiedName {
```
Question: why remove QualifiedNamePtr?

suo replied:
I explain the reasoning in the PR description

eellison reviewed (approved):
Looks fine, but now each Qualified name is taking up twice as much storage, prefix + name_ and qualifiedName_.

suo committed:
Update on "[jit] make QualifiedName a value type"

[jit] make QualifiedName a value type

Previously QualifiedName was a recursive reference type, which is more
efficient as it avoids copies. But I want to make it a value type
because:
- People think of names as values anyway
- When doing stuff with QualifiedName I found myself doing a lot of
recursive prefix traversals and copies anyway.

gh-metadata: pytorch pytorch 19626 gh/suo/31/head
